addon/services/file-queue.ts

```diff
@@ -0,0 +1,126 @@
+import { assert } from '@ember/debug';
+import Service from '@ember/service';
+import { registerDestructor } from '@ember/destroyable';
+
+import Queue, { QueueName } from '../queue';
+import File from '../file';
+
+export const DEFAULT_QUEUE = Symbol('DEFAULT_QUEUE');
+
+/**
+ * The file queue service is a global file
+ * queue that manages all files being uploaded.
+ *
+ * This service can be used to query the current
+ * upload state when a user leaves the app,
+ * asking them whether they want to cancel
+ * the remaining uploads.
+ */
+export default class FileQueueService extends Service {
+  queues: Map<QueueName, Queue> = new Map();
+
+  /**
+   * Returns a queue with the given name
+   *
+   * @param name The name of the queue to find
+   * @returns The queue if it exists
+   */
+  find(name: QueueName): Queue | undefined {
+    return this.queues.get(name);
+  }
+
+  /**
+   * Create a new queue with the given name.
+   *
+   * @param name The name of the queue to create
+   * @returns The new queue.
+   */
+  create(name: QueueName): Queue {
+    assert(
+      `Queue names are required to be unique. "${String(
+        name
+      )}" has already been reserved.`,
+      !this.queues.has(name)
+    );
+
+    const queue = new Queue({ name, fileQueue: this });
+
+    registerDestructor(queue, () => {
+      this.queues.delete(name);
+    });
+
+    this.queues.set(name, queue);
+    return queue;
+  }
+
+  findOrCreate(name: QueueName): Queue {
+    return this.find(name) ?? this.create(name);
+  }
+
+  //
+  // @TODO
+  // Everything below this line should be deprecated ?
```

Comment:
I guess this API is confusing since there are some shared properties with `Queue` right? Not sure how much they are used by addon consumers.

Reply:
Deprecate, because they are part of public API and potentially are used as part of this "super-global" queue. They should receive deprecation warnings 👍